### PR TITLE
TracAlgorithm

### PR DESCRIPTION
**Pull request by emailweixu**
It compares the action distributions after the SGD with the action
    distributions from the previous model. If the KL-divergence is too big,
    the new parameters are shrinked as:
        w_new' = old_w + max_kl / kl * (w_new - w_old)

### CODE REVIEW

File: alf/utils/math_ops.py [hunk 1]
```diff
@@ -53,3 +53,31 @@ def _ignore(t):
...
+
+def max_n(inputs):
+    """Calculate the maximum of n Tensors
+
+    Args:
+        inputs (list[Tensor]): list of Tensors, should have the shape
+    Returns:
+        the elementwise maximum of all the tensors in `inputs`
+    """
+    ret = inputs.pop()
```
Comment (witwolf) [inline alf/utils/math_ops.py]: perharps  `ret = inputs[0]`  is better, do not modify input arguments

File: alf/utils/math_ops.py [hunk 2]
```diff
@@ -53,3 +53,31 @@ def _ignore(t):
...
+        Tensor
+    """
+    return x * tf.math.sigmoid(x)
+
+
+def max_n(inputs):
+    """Calculate the maximum of n Tensors
+
+    Args:
+        inputs (list[Tensor]): list of Tensors, should have the shape
```
Comment (witwolf) [inline alf/utils/math_ops.py]: should have the shape => should have same shape

File: alf/algorithms/trac_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        # repeated if the final_step_mode of OnPolicyDriver is FINAL_STEP_REDO.
+        # See comment of OnPolicyDriver for detail.
+        self._buffer.pop(self._batch_size)
+        size = self._buffer.current_size
+        total_kls = nest_map(lambda _: tf.zeros(()), self.action_spec)
+        for t in tf.range(0, size, self._batch_size):
+            exp = self._buffer.get_batch_by_indices(
+                tf.range(t, t + self._batch_size))
+            # Here we use stored states to calculate action. It might be better
+            # to use the states calculated using the new weights.
```
Comment (hnyu) [inline alf/algorithms/trac_algorithm.py]: Might have problems with RNN-actor networks.

Comment (emailweixu) [reply alf/algorithms/trac_algorithm.py]: It's not clear which is good for RNN. It's relatively easy to use the RNN state if we want.

File: alf/algorithms/trac_algorithm.py [hunk 4]
```diff
@@ -12,13 +12,16 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """Trusted Region Actor critic algorithm."""
+import numbers
```
Comment (hnyu) [inline alf/algorithms/trac_algorithm.py]: not used

File: alf/algorithms/trac_algorithm.py [hunk 5]
```diff
@@ -135,13 +146,13 @@ def rollout(self, time_step: ActionTimeStep, state: ActorCriticState):
...
-        kls, steps = self._trusted_updater.adjust_step(self._calc_change,
-                                                       self._action_dist_clips)
+        dists, steps = self._trusted_updater.adjust_step(
+            self._calc_change, self._action_dist_clips)
 
         def _summarize():
             with self.name_scope:
-                for i, kl in enumerate(tf.nest.flatten(kls)):
-                    tf.summary.scalar("action_kl/%s" % i, kl)
+                for i, kl in enumerate(tf.nest.flatten(dists)):
```
Comment (hnyu) [inline alf/algorithms/trac_algorithm.py]: kl -> d

File: alf/algorithms/trac_algorithm.py [hunk 6]
```diff
@@ -135,13 +146,13 @@ def rollout(self, time_step: ActionTimeStep, state: ActorCriticState):
...
+                    tf.summary.scalar("action_dist/%s" % i, kl)
```
Comment (hnyu) [inline alf/algorithms/trac_algorithm.py]: action_dist -> unadjusted_action_distance

**Pull request merged**